% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
  12pt,
]{article}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newlength{\cslentryspacingunit} % times entry-spacing
\setlength{\cslentryspacingunit}{\parskip}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
  \let\oldpar\par
  \def\par{\hangindent=\cslhangindent\oldpar}
  \fi
  % set entry spacing
  \setlength{\parskip}{#2\cslentryspacingunit}
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}
\usepackage{booktabs}
\usepackage{longtable}
\usepackage{array}
\usepackage{multirow}
\usepackage{wrapfig}
\usepackage{float}
\usepackage{colortbl}
\usepackage{pdflscape}
\usepackage{tabu}
\usepackage{threeparttable}
\usepackage{threeparttablex}
\usepackage[normalem]{ulem}
\usepackage{makecell}
\usepackage{xcolor}
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
\hypersetup{
  pdftitle={Phylogenetic networks for detecting ancient hybridization},
  pdfauthor={Carolina Siniscalchi, Erika Moore-Pollard, Paige Ellestad, \& Jennifer Mandel},
  colorlinks=true,
  linkcolor={Maroon},
  filecolor={Maroon},
  citecolor={Blue},
  urlcolor={blue},
  pdfcreator={LaTeX via pandoc}}

\title{Phylogenetic networks for detecting ancient hybridization}
\usepackage{etoolbox}
\makeatletter
\providecommand{\subtitle}[1]{% add subtitle to \maketitle
  \apptocmd{\@title}{\par {\large #1 \par}}{}{}
}
\makeatother
\subtitle{Bioinformatic pipeline}
\author{Carolina Siniscalchi, Erika Moore-Pollard, Paige Ellestad, \& Jennifer Mandel}
\date{2025-01-05}

\begin{document}
\maketitle

{
\hypersetup{linkcolor=}
\setcounter{tocdepth}{2}
\tableofcontents
}
\href{https://github.com/svenbuerki/VanillaGenomicsCode}{Raw data on GitHub}

\hypertarget{introduction}{%
\section{Introduction}\label{introduction}}

an intro

\hypertarget{sampling}{%
\section{Sampling}\label{sampling}}

sampling

\hypertarget{distribution-data}{%
\subsection{Distribution data}\label{distribution-data}}

\hypertarget{genomic-data}{%
\subsection{Genomic data}\label{genomic-data}}

\hypertarget{overview-of-workflow}{%
\section{Overview of workflow}\label{overview-of-workflow}}

This section provides an overview of the analyses conducted in this study aiming at comparing the genome complexity and structure of samples of \emph{Vanilla planifolia} from Mexico and data available on NCBI (see Table \ref{tab:tabsampling}). Details on the software and associated code for the following analyses are presented in this document:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  \protect\hyperlink{setenv}{Setting working environment}.
\item
  \protect\hyperlink{dwndata}{Downloading SRA files}.
\item
  \protect\hyperlink{trim}{Conducting reads trimming}.
\item
  \protect\hyperlink{GS}{Genome size and heterozygosity}.
\item
  \protect\hyperlink{cyt}{Ploidy level and genome complexity}.
\item
  \protect\hyperlink{conseq}{Reconstructing genome sequences}.
\item
  \protect\hyperlink{comgen}{Comparative chromosome level analyses}.
\end{enumerate}

\hypertarget{references}{%
\section{References}\label{references}}

\hypertarget{refs}{}
\begin{CSLReferences}{0}{0}
\end{CSLReferences}

\hypertarget{appendix-1}{%
\section{Appendix 1}\label{appendix-1}}

Citations of all R packages used to generate this report.

{[}1{]} J. Allaire, Y. Xie, C. Dervieux, et al.~\emph{rmarkdown: Dynamic
Documents for R}. R package version 2.29. 2024.
\url{https://github.com/rstudio/rmarkdown}.

{[}2{]} H. Bengtsson. \emph{R.methodsS3: S3 Methods Simplified}. R package
version 1.8.2. 2022. \url{https://github.com/HenrikBengtsson/R.methodsS3}.

{[}3{]} H. Bengtsson. \emph{R.oo: R Object-Oriented Programming with or without
References}. R package version 1.27.0. 2024.
\url{https://github.com/HenrikBengtsson/R.oo}.

{[}4{]} H. Bengtsson. \emph{R.utils: Various Programming Utilities}. R package
version 2.12.3. 2023. \url{https://henrikbengtsson.github.io/R.utils/}.

{[}5{]} H. Bengtsson. ``The R.oo package - Object-Oriented Programming with
References Using Standard R Code''. In: \emph{Proceedings of the 3rd
International Workshop on Distributed Statistical Computing (DSC
2003)}. Ed. by K. Hornik, F. Leisch and A. Zeileis. Vienna, Austria,
Mar.~2003.
\url{https://www.r-project.org/conferences/DSC-2003/Proceedings/Bengtsson.pdf}.

{[}6{]} H. Bengtsson. ``The R.oo package - Object-Oriented Programming with
References Using Standard R Code''. In: \emph{Proceedings of the 3rd
International Workshop on Distributed Statistical Computing (DSC
2003)}. Ed. by K. Hornik, F. Leisch and A. Zeileis. Vienna, Austria,
Mar.~2003.
\url{https://www.r-project.org/conferences/DSC-2003/Proceedings/Bengtsson.pdf}.

{[}7{]} C. Boettiger. \emph{knitcitations: Citations for Knitr Markdown Files}.
R package version 1.0.12. 2021.
\url{https://github.com/cboettig/knitcitations}.

{[}8{]} D. Charif and J. Lobry. ``SeqinR 1.0-2: a contributed package to the
R project for statistical computing devoted to biological sequences
retrieval and analysis.'' In: \emph{Structural approaches to sequence
evolution: Molecules, networks, populations}. Ed. by U. Bastolla, M.
Porto, H. Roman and M. Vendruscolo. Biological and Medical Physics,
Biomedical Engineering. ISBN : 978-3-540-35305-8. New York: Springer
Verlag, 2007, pp.~207-232.

{[}9{]} D. Charif and J. R. Lobry. \emph{seqinr: Biological Sequences Retrieval
and Analysis}. R package version 4.2-36. 2023.
\url{https://seqinr.r-forge.r-project.org/}.

{[}10{]} J. Cheng, B. Schloerke, B. Karambelkar, et al.~\emph{leaflet: Create
Interactive Web Maps with the JavaScript Leaflet Library}. R package
version 2.2.2. 2024. \url{https://rstudio.github.io/leaflet/}.

{[}11{]} M. C. Koohafkan. \emph{kfigr: Integrated Code Chunk Anchoring and
Referencing for R Markdown Documents}. R package version 1.2.1. 2021.
\url{https://github.com/mkoohafkan/kfigr}.

{[}12{]} J. Ooms. \emph{magick: Advanced Graphics and Image-Processing in R}. R
package version 2.8.5. 2024. \url{https://docs.ropensci.org/magick/}.

{[}13{]} R Core Team. \emph{R: A Language and Environment for Statistical
Computing}. R Foundation for Statistical Computing. Vienna, Austria,
2023. \url{https://www.R-project.org/}.

{[}14{]} H. Wickham. \emph{ggplot2: Elegant Graphics for Data Analysis}.
Springer-Verlag New York, 2016. ISBN: 978-3-319-24277-4.
\url{https://ggplot2.tidyverse.org}.

{[}15{]} H. Wickham, J. Bryan, M. Barrett, et al.~\emph{usethis: Automate
Package and Project Setup}. R package version 3.1.0. 2024.
\url{https://usethis.r-lib.org}.

{[}16{]} H. Wickham, W. Chang, L. Henry, et al.~\emph{ggplot2: Create Elegant
Data Visualisations Using the Grammar of Graphics}. R package version
3.5.1. 2024. \url{https://ggplot2.tidyverse.org}.

{[}17{]} H. Wickham, R. Fran√ßois, L. Henry, et al.~\emph{dplyr: A Grammar of
Data Manipulation}. R package version 1.1.4. 2023.
\url{https://dplyr.tidyverse.org}.

{[}18{]} H. Wickham, J. Hester, W. Chang, et al.~\emph{devtools: Tools to Make
Developing R Packages Easier}. R package version 2.4.5. 2022.
\url{https://devtools.r-lib.org/}.

{[}19{]} D. Winter. \emph{pafr: Read, Manipulate and Visualize Pairwise mApping
Format Data}. R package version 0.0.2. 2020.
\url{https://dwinter.github.io/pafr/}.

{[}20{]} Y. Xie. \emph{bookdown: Authoring Books and Technical Documents with R
Markdown}. Boca Raton, Florida: Chapman and Hall/CRC, 2016. ISBN:
978-1138700109. \url{https://bookdown.org/yihui/bookdown}.

{[}21{]} Y. Xie. \emph{bookdown: Authoring Books and Technical Documents with R
Markdown}. R package version 0.41. 2024.
\url{https://github.com/rstudio/bookdown}.

{[}22{]} Y. Xie. \emph{Dynamic Documents with R and knitr}. 2nd. ISBN
978-1498716963. Boca Raton, Florida: Chapman and Hall/CRC, 2015.
\url{https://yihui.org/knitr/}.

{[}23{]} Y. Xie. \emph{formatR: Format R Code Automatically}. R package version
1.14. 2023. \url{https://github.com/yihui/formatR}.

{[}24{]} Y. Xie. ``knitr: A Comprehensive Tool for Reproducible Research in
R''. In: \emph{Implementing Reproducible Computational Research}. Ed. by V.
Stodden, F. Leisch and R. D. Peng. ISBN 978-1466561595. Chapman and
Hall/CRC, 2014.

{[}25{]} Y. Xie. \emph{knitr: A General-Purpose Package for Dynamic Report
Generation in R}. R package version 1.49. 2024.
\url{https://yihui.org/knitr/}.

{[}26{]} Y. Xie and J. Allaire. \emph{tufte: Tufte's Styles for R Markdown
Documents}. R package version 0.13. 2023.
\url{https://github.com/rstudio/tufte}.

{[}27{]} Y. Xie, J. Allaire, and G. Grolemund. \emph{R Markdown: The Definitive
Guide}. Boca Raton, Florida: Chapman and Hall/CRC, 2018. ISBN:
9781138359338. \url{https://bookdown.org/yihui/rmarkdown}.

{[}28{]} Y. Xie, C. Dervieux, and E. Riederer. \emph{R Markdown Cookbook}. Boca
Raton, Florida: Chapman and Hall/CRC, 2020. ISBN: 9780367563837.
\url{https://bookdown.org/yihui/rmarkdown-cookbook}.

{[}29{]} H. Zhu. \emph{kableExtra: Construct Complex Table with kable and Pipe
Syntax}. R package version 1.4.0. 2024.
\url{http://haozhu233.github.io/kableExtra/}.

\end{document}
